File: Data Structures & Algorithms/binary-tree-right-side-view/submission-2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type Queue struct {
	queue []*TreeNode
}

func (q *Queue) enqueue(el *TreeNode) {
	q.queue = append(q.queue, el)
}

func (q *Queue) dequeue() *TreeNode {
	el := q.queue[0]
	q.queue = q.queue[1:]
	return el
}

func (q *Queue) length() int {
	return len(q.queue)
}

func rightSideView(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}

	queue := Queue{
		queue: []*TreeNode{root},
	}

	result := []int{}

	for queue.length() > 0 {
		queueLength := queue.length()

		for i := range queueLength {
			node := queue.dequeue()

			if node.Left != nil {
				queue.enqueue(node.Left)
			}

			if node.Right != nil {
				queue.enqueue(node.Right)
			}

			if i == queueLength-1 {
				result = append(result, node.Val)
			}

		}
	}

	return result
}

File: Data Structures & Algorithms/delete-node-in-a-bst/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  public TreeNode deleteNode(TreeNode root, int key) {
    if(root == null){
      return null;
    }

    if (root.val == key) {
      if (root.left == null && root.right == null) {
        return null;
      } else if (root.left == null) {
        return root.right;
      } else if (root.right == null) {
        return root.left;
      } else {
        TreeNode min = findMin(root.right);
        root.right = deleteNode(root.right, min.val);
        root.val = min.val;

        return root;
      }

    } else if (root.val > key) {
      root.left = deleteNode(root.left, key);
      return root;
    } else {
      root.right = deleteNode(root.right, key);
      return root;
    }
  }

  private TreeNode findMin(TreeNode root) {
    if (root.left == null) {
      return root;
    }

    return findMin(root.left);
  }
}